src/compiler/crystal/codegen/cache_dir.cr

```diff
@@ -0,0 +1,63 @@
+module Crystal
+  # Manages cache files in the ".crystal" directory.
+  #
+  # For each compiled program a directory is created in the cache
+  # that stores .bc and .o files that could possibly be reused
+  # from a previous compilation.
+  #
+  # To keep the cache dir small, only the 10 most recently used
+  # directories are kept. We use the directory's modification
+  # time for this.
+  class CacheDir
+    def initialize(@dir : String)
+    end
+
+    # Returns the directory where cache files related to the
+    # given sources will be stored. The directory will be
+    # created if it doesn't exist.
+    def directory_for(sources : Array(Compiler::Source))
+      directory_for(sources.first.filename)
+    end
+
+    # Returns the directory where cache files related to the
+    # given filenames will be stored. The directory will be
+    # created if it doesn't exist.
+    def directory_for(filename : String)
+      name = filename.gsub('/', '-')
+      while name.starts_with?('-')
+        name = name[1..-1]
+      end
+      output_dir = File.join(@dir, name)
+      Dir.mkdir_p(output_dir)
+      output_dir
+    end
+
+    # Keeps the 10 most recently used directories in the cache,
+    # and removes all others. This also removes non-directory
+    # files inside the cache directory (temporary executables
+    # resulting from `crystal run` or `run` macro calls).
+    def cleanup
+      cleanup_dirs
+      cleanup_files
+    end
+
+    private def cleanup_dirs
+      Dir.entries(@dir)
+         .reject { |name| name == "." || name == ".." }
+         .map! { |name| File.join(@dir, name) }
+         .select { |dir| Dir.exists?(dir) }
+         .sort_by! { |dir| File.stat(dir).mtime rescue Time.epoch(0) }
+         .reverse!
+         .skip(10)
+         .each { |name| `rm -rf "#{name}"` rescue nil }
```

Comment:
I actually searched if we have a method like that before writing this. I couldn't find it so I thought "maybe we are using `rm -rf ...`?" and I was right. When we add `Dir.delete(path, recursive: true)` I'm sure we will remember changing all those `rm -rf ...`. Or it can be added as a reminder for the issue that tracks that feature.